### PR TITLE
Plugins: Support for Train Plugins in InSpec

### PR DESCRIPTION
**Pull request by clintoncwolfe**
This PR adds support for managing Train transport plugins in InSpec.

A gem, `train-test-fixture` now exists on rubygems.org . That gem is managed in the Train repo, PR pending.

 * Docs updated to mention Train plugins
 * Certain UX messaging updated to mention Train
 * Small modifications to the plugin loader to accommodate Train plugins
 * Functional tests updated to verify ability to search for and install train plugins
 * Functional tests updated to verify that `inspec detect` and `inspec shell` work with the test fixture plugin.

Do not be alarmed by the high line-change count; about 800 of that is in the installed copy of the `train-test-fixture` gem, which is not being managed (and need not be reviewed) here.

### CODE REVIEW

**Review by jquick — APPROVED** [review]
Looks good to me! Thanks for your work on this @clintoncwolfe

File: docs/plugins.md [hunk 1]
```diff
@@ -1,22 +1,26 @@
...
+title: About InSpec and Train Plugins
 ---
 
-# InSpec Plugins
+# InSpec and Train Plugins
 
 ## What are InSpec Plugins?
 
-Plugins are optional software components that extend the capabilities of InSpec. For example, `inspec-iggy` is a plugin project that aims to generate InSpec controls from infrastructure-as-code.  Plugins are distributed as RubyGems, and InSpec manages their installation.
+InSpec Plugins are optional software components that extend the capabilities of InSpec. For example, `inspec-iggy` is a plugin project that aims to generate InSpec controls from infrastructure-as-code.  Plugins are distributed as RubyGems, and InSpec manages their installation. InSpec plugins always begin with the prefix 'inspec-'.
```
Comment (jerryaldrichiii) [inline docs/plugins.md]: Should we link to iggy? I'd also rephrase it as `controls from infrastructre-as-code state files`.

Comment (jerryaldrichiii) [reply docs/plugins.md]: Thoughts on `must begin` vs `always begin`

Comment (jerryaldrichiii) [inline docs/plugins.md]: Seems we use `InSpec Plugins` and `InSpec plugins` throughout. I think we should standardize.

File: docs/plugins.md [hunk 2]
```diff
@@ -1,22 +1,26 @@
...
+
+## What are Train Plugins?
+
+Train plugins allow InSpec to speak to new kinds of remote targets.  For example, if you wanted to audit a Kubernetes cluster, you might want a transport that can talk to the supervisor API.  You'd develop a train plugin for that, and install it using the InSpec command line. Train plugins always begin with the prefix 'train-'.
```
Comment (jerryaldrichiii) [inline docs/plugins.md]: Always remote targets? Could there be local targets? Example: Interfacing with a binary on the local system.

Also, I think `train plugin` should be `Train plugin`

Comment (jerryaldrichiii) [reply docs/plugins.md]: ~Same as above `always` vs `must`?~ Covered above.

File: test/functional/plugins_test.rb [hunk 3]
```diff
@@ -104,3 +104,59 @@
...
+      lines = outcome.stdout.split("\n")
+      lines.grep(/Name/).first.must_include('test-fixture')
+      lines.grep(/Name/).first.wont_include('train-test-fixture')
+      lines.grep(/Release/).first.must_include('0.1.0')
+      lines.grep(/Families/).first.must_include('os')
+      lines.grep(/Families/).first.must_include('windows')
+      lines.grep(/Families/).first.must_include('unix')
+      lines.grep(/Arch/).first.must_include('mock')
+    end
+    it 'can run inspec detect against a test-fixture backend' do
```
Comment (jerryaldrichiii) [inline test/functional/plugins_test.rb:129]: Other `it` blocks have spaces between them, could we add one here?

File: test/functional/plugins_test.rb [hunk 4]
```diff
@@ -104,3 +104,59 @@
...
+      outcome.exit_status.must_equal(0)
+      outcome.stderr.must_be_empty
+      outcome.stdout.chomp.must_equal "17"
+
+      outcome = inspec_with_env("shell -t test-fixture:// -c 'command(\"echo hello\").stdout'",  INSPEC_CONFIG_DIR: File.join(config_dir_path, 'train-test-fixture'))
+      outcome.exit_status.must_equal(0)
+      outcome.stderr.must_be_empty
+      outcome.stdout.chomp.must_equal "Mock Command Result stdout"
+    end
+
```
Comment (jerryaldrichiii) [inline test/functional/plugins_test.rb]: Extra newline...but alas no RuboCop. Ignore at will.

File: test/unit/mock/config_dirs/train-test-fixture/gems/2.3.0/gems/train-test-fixture-0.1.0/lib/train-test-fixture/connection.rb [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        )
+      end
+
+      def file_via_connection(path, *args)
+        MockFile.new(self, path)
+      end
+
+      class MockFile < Train::File
+        def content
+          # Remarkably, the content is always the same.
```
Comment (jerryaldrichiii) [inline test/unit/mock/config_dirs/train-test-fixture/gems/2.3.0/gems/train-test-fixture-0.1.0/lib/train-test-fixture/connection.rb:33]: Heh, reading your work always gives me a chuckle.

**Review by jerryaldrichiii — APPROVED** [review]
Looks good to me. Only a few questions/comments.

~Where can I review the `train-test-fixture` gem?...I'm seeing quite a few things (mostly spelling/grammar things) that I'd like to highlight.~ ~EDIT: Disregard...it was next on my list.~ EDIT EDIT: My shame is unbearable...I'm not sure where to air my concerns about `train-test-fixture`.

**Comment by clintoncwolfe** [comment]
Tagged DNM; @chris-rock reported installing train plugins by path ref fails on load probe, fix in hand & will push within the hour
EDIT: Removed DNM, fixed on e2502b4

File: docs/plugins.md (hunk 1 above)
Comment (clintoncwolfe) [reply docs/plugins.md]: I'd say always.  The audience for this doc is plugin *users*, who will want to know how to identify a gem they could install.  A plugin *developer* on the other hand will need to know that they *must* name their gem a certain way.

**Comment by clintoncwolfe** [comment]
@jerryaldrichiii as noted in the description of the PR, train-test-fixture is not in scope for this review. Please do not spend time on it in this PR.

The code present on this PR is the result of a manual install of the gem.  The gem is available on rubygems.org.  The code here is mere present so we can test `inspec plugin list` without attempting an install.

The code is developed on https://github.com/inspec/train/pull/356 , which you have already reviewed.

**Review by jerryaldrichiii — APPROVED** [review]

**Pull request merged**